### DISCUSSION TITLE
Control to slide between pages with the mouse?

### DISCUSSION THREAD

**Discussion by mysteryx93** [AvaloniaUI/Avalonia#14203] [General]
In mobile apps, you often slide between pages. Is there a control in Avalonia that's like a Tab Control or Caroussel that you swipe with the mouse?

**Comment by thevortexcloud** [comment]
Avalonia literally has a Carousel control.

https://docs.avaloniaui.net/docs/reference/controls/carousel

Not sure if it lets you swipe though.

**Reply by mysteryx93** [reply]
From the documentation it doesn't seem it supports swipe, that was my first hope

Would need both a caroussel that swipes, and also a TabControl that swipes between tabs

**Reply by thevortexcloud** [reply]
If you don't need to smoothly swipe you can probably make your own fairly easily by adding a gesture recognizer.

https://docs.avaloniaui.net/docs/reference/gestures/scrollgesturerecognizer

**Reply by robloo** [reply]
@mysteryx93 Carousel is being updated and should work with swipe: https://github.com/AvaloniaUI/Avalonia/pull/20659

I would recommend just using this in-box control as soon as that PR is in for 12.0 previews.

**Comment by mysteryx93** [comment]
What we'd need is TabbedPage, SwipeView and various swipe action controls
https://learn.microsoft.com/en-us/xamarin/xamarin-forms/platform/android/tabbedpage-page-swiping
https://learn.microsoft.com/en-us/xamarin/xamarin-forms/user-interface/swipeview

There's this threat that was discussing the issue
https://github.com/AvaloniaUI/Avalonia/issues/10557

Until then; it's very difficult to develop any real mobile app. It seems like everyone is just waiting for everyone to develop the base components before doing any mobile development.

**Reply by workgroupengineering** [reply]
see https://github.com/AvaloniaUI/Avalonia.Labs

**Reply by tkefauver** [reply]
It is difficult. [Material.Avalonia](https://github.com/AvaloniaCommunity/Material.Avalonia) helps for some things. Like the ripple effect and card views for me at least.

**Reply by stevemonaco** [reply]
I have a sample at https://github.com/stevemonaco/AvaloniaDemos/tree/master/SwipeNavigation which lays most of the groundwork for a swipe-based page navigator. Requires manual creation/caching of pages from code-behind and doesn't have things like `ItemsSource`. I didn't really test for pages being added/removed, so it's not production-ready.

**Reply by mysteryx93** [reply]
by the way I implemented manual creation/caching of pages in MvvmDialogs with WeakReferences but the weak links aren't working, not allowing the cache to invalidate if memory is needed. Never had the time to look further into it. If someone can find the issue, that would be great. Perhaps something similar could help with SwipeNavigation.
https://github.com/mysteryx93/HanumanInstitute.MvvmDialogs/blob/master/src/MvvmDialogs.Avalonia/Navigation/NavigationManager.cs

**Comment by davomelkumyan40** [comment]
Check Avalonia.Labs

**Reply by robloo** [reply]
Note Carousel is being updated and should support the needed cases going forward: https://github.com/AvaloniaUI/Avalonia/pull/20659
